File: ACTIVIDADES/Actividad N°5/Insertion_Sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>
#include <random>

// Algoritmo Insertion Sort
void insertion_sort(std::vector<int>& arr) {
    for (size_t i = 1; i < arr.size(); ++i) {
        int key = arr[i];
        int j = i - 1;
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            --j;
        }
        arr[j + 1] = key;

        if (i < 6) { // Mostrar primeras 5 iteraciones
            std::cout << "Iteracion " << i << ": ";
            for (size_t k = 0; k < std::min(arr.size(), size_t(10)); ++k) {
                std::cout << arr[k] << " ";
            }
            std::cout << "...\n";
        }
    }
}

// Función para probar el algoritmo y medir tiempo
void probar_insertion_sort(const std::string& nombre, const std::vector<int>& arreglo) {
    std::cout << "\n--- " << nombre << " ---\n";
    std::vector<int> copia = arreglo;
    auto inicio = std::chrono::high_resolution_clock::now();
    insertion_sort(copia);
    auto fin = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> duracion = fin - inicio;
    std::cout << nombre << " - Tiempo: " << duracion.count() << " segundos\n";
}

int main() {
    const int n = 1000;
    
    // Crear vector aleatorio
    std::vector<int> aleatorio(n);
    for (int i = 0; i < n; ++i) {
        aleatorio[i] = i;
    }
    std::random_device rd;
    std::mt19937 g(rd());
    std::shuffle(aleatorio.begin(), aleatorio.end(), g);

    // Crear vector ordenado
    std::vector<int> ordenado(n);
    for (int i = 0; i < n; ++i) {
        ordenado[i] = i;
    }

    // Crear vector inverso
    std::vector<int> inverso(n);
    for (int i = 0; i < n; ++i) {
        inverso[i] = n - i;
    }

    std::cout << "Insertion Sort:\n";
    probar_insertion_sort("Aleatorio", aleatorio);
    probar_insertion_sort("Ordenado", ordenado);
    probar_insertion_sort("Inverso", inverso);

    return 0;
}

File: ACTIVIDADES/Actividad N°5/insertion_sort.py
import random
import time

# Algoritmo Insertion Sort
def insertion_sort(arr):
    for i in range(1, len(arr)):  # Comenzar desde el segundo elemento
        key = arr[i]  # Valor actual a insertar
        j = i - 1  # Comparar con elementos anteriores
        while j >= 0 and arr[j] > key:  # Mientras haya elementos mayores
            arr[j + 1] = arr[j]  # Desplazar a la derecha
            j -= 1
        arr[j + 1] = key  # Insertar en su lugar correcto
        if i < 6:  # Mostrar primeras 5 iteraciones
            print(f"Iteración {i}: {arr[:10]}...")  # Muestra parcial

# Función para probar el algoritmo y medir tiempo
def probar_insertion_sort(nombre, arreglo):
    print(f"\n--- {nombre.upper()} ---")
    copia = arreglo.copy()
    inicio = time.time()
    insertion_sort(copia)
    fin = time.time()
    print(f"{nombre} - Tiempo: {fin - inicio:.6f} segundos")

# Generación de arreglos
n = 1000
aleatorio = random.sample(range(n), n)
ordenado = list(range(n))
inverso = list(range(n, 0, -1))

# Ejecución
print("Insertion Sort:")
probar_insertion_sort("Aleatorio", aleatorio)
probar_insertion_sort("Ordenado", ordenado)
probar_insertion_sort("Inverso", inverso)
